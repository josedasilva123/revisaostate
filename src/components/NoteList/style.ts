import styled from "styled-components";

export const StyledNoteList = styled.ul`
    display: flex;
    flex-direction: column;
    gap: 2rem;
   
`