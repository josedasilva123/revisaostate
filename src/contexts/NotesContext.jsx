import { createContext } from "react";
import { useState } from "react";

export const NotesContext = createContext({});

export const NotesProvider = ({ children }) => {
  const noteList = [
    {
      id: 0,
      title: "Nota 1",
      text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. In in placerat orci. Vestibulum sit amet dapibus elit. Nullam bibendum nulla in ex elementum convallis.",
    },
    {
      id: 1,
      title: "Nota 2",
      text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. In in placerat orci. Vestibulum sit amet dapibus elit. Nullam bibendum nulla in ex elementum convallis.",
    },
    {
      id: 2,
      title: "Nota 3",
      text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. In in placerat orci. Vestibulum sit amet dapibus elit. Nullam bibendum nulla in ex elementum convallis.",
    },
  ];

  const [notes, setNotes] = useState(noteList);
  const [currentNote, setCurrentNote] = useState(null); 
  const [counter, setCounter] = useState(noteList.length);

  function createNote(formData) {
    const newNote = {
      id: counter,
      title: formData.title,
      text: formData.text,
    };

    setNotes([...notes, newNote]);

    setCounter(counter + 1);
  }

  function removeNote(id) {
    const newList = notes.filter((note) => note.id !== id);
    setNotes(newList);
  }

  function editNote(formData){   
    const noteIndex = notes.indexOf(currentNote);   
    const newNotes = [...notes];
    
    newNotes[noteIndex].text = formData.text;

    setNotes(newNotes);    
    setCurrentNote(null);
  }

  /* Export */
  return (
    <NotesContext.Provider value={{ notes, currentNote, setCurrentNote, createNote, removeNote, editNote }}>
      {children}
    </NotesContext.Provider>
  );
};
