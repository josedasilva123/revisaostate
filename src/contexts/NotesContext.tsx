import { AxiosError } from "axios";
import React, { useState, useEffect, createContext, useContext } from "react";
import { SubmitHandler } from "react-hook-form";
import { toast } from "react-toastify"
import { api } from "../api/api";
import { UserContext } from "./UserContext";

export interface IFormCreateNote{
  title: string;
  text: string;
}

export interface IFormEditNote{
  title: string;
  text: string;
}

export interface INote{
  _id: string;
  userID: string;
  title: string;
  text:  string;
  createdAt: string;
  updatedAt: string; 
  __v: number;
}

interface INotesContext{
  notes: INote[];
  currentNote: INote | null;
  setCurrentNote: React.Dispatch<React.SetStateAction<INote | null>>;
  createNote: (formData: IFormCreateNote, setLoading: React.Dispatch<React.SetStateAction<boolean>>) => void;
  removeNote: (id: string) => void;
  editNote: SubmitHandler<IFormEditNote>;
}

interface INotesProps{
  children: React.ReactNode;
}

interface IError{
  error: string;
}

export const NotesContext = createContext<INotesContext>({} as INotesContext);

export const NotesProvider = ({ children }: INotesProps) => { 
  const { user } = useContext(UserContext);
  const [notes, setNotes] = useState<INote[]>([] as INote[]);
  const [currentNote, setCurrentNote] = useState<INote | null>(null); 

  //Nesta api é necessário uma requisição
  useEffect(() => {
    const token = localStorage.getItem('@TOKEN');

    async function getNotes(){
      try {        
        const response = await api.get('notes', {
          headers: {
            auth: token as string,
          }
        })        
        setNotes(response.data.response);
      } catch (error) {
        console.log(error);
      }
    }

    //Executa função somente com usuário válido e token
    if(user && token){
      getNotes();
    } else {
      //Do contrario limpa notas
      setNotes([]);
    }     

  }, [user]);

  async function createNote(formData: IFormCreateNote, setLoading: React.Dispatch<React.SetStateAction<boolean>>){
    try {
      setLoading(true);
      const token = localStorage.getItem('@TOKEN');
      const response = await api.post('notes', formData, {
        headers: {
          auth: token as string,
        }
      })
      //Mesmo com a requisição a renderização na interface 
      setNotes([...notes, response.data.response]); 
      toast.success(response.data.message);
    } catch (error) {
      const serverError = error as AxiosError<IError>;
      toast.error(serverError.response?.data.error);
    } finally {
      setLoading(false);
    }
  }

  async function removeNote(id: string) {
    try {
      const token = localStorage.getItem('@TOKEN');
      const response = await api.delete(`notes/${id}`, {
        headers: {
          auth: token as string,
        }
      })
      toast.success(response.data.message);
      const newList = notes.filter((note) => note._id !== id);
      setNotes(newList);
    } catch (error) {
      const serverError = error as AxiosError<IError>;
      toast.error(serverError.response?.data.error);
    }
    /*
    const newList = notes.filter((note) => note.id !== id);
    setNotes(newList);
    toast.success('Nota removida com sucesso!')
    */
  }
  
  const editNote: SubmitHandler<IFormEditNote> = (formData) => {
    const noteIndex = notes.indexOf(currentNote as INote);
    
    const newNotes = [...notes];
    
    newNotes[noteIndex].text = formData.text;
    
    setNotes(newNotes);    
    setCurrentNote(null);
    toast.success('Nota alterada com sucesso!')
  }

  /* Export */
  return (
    <NotesContext.Provider value={{ notes, currentNote, setCurrentNote, createNote, removeNote, editNote }}>
      {children}
    </NotesContext.Provider>
  );
};
