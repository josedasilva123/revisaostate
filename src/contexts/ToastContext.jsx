/* eslint-disable react-hooks/exhaustive-deps */
import { createContext, useState, useEffect, useCallback } from "react";

export const ToastContext = createContext({});

export const ToastProvider = ({ children }) => {
  const [toastList, setToastList] = useState([]);
  const [counter, setCounter] = useState(1);  

  const deleteToast = useCallback(
    (id) => {
      const newToastList = toastList.filter((e) => e.id !== id);
      setToastList(newToastList);
    },
    [toastList, setToastList]
  );

  useEffect(() => {
    const interval = setInterval(() => {
      if (toastList.length) {
        deleteToast(toastList[0].id);
      }
    }, 3000);
    return () => {
      clearInterval(interval);
    };
  }, [toastList, deleteToast]);

  function createToast({ message, type, duration }) {
    const newToast = {
      id: counter,
      message,
      type,
    };
    setToastList([...toastList, newToast]);
    setCounter(counter + 1);
  }

  return (
    <ToastContext.Provider value={{ toastList, createToast, deleteToast }}>
      {children}
    </ToastContext.Provider>
  );
};
