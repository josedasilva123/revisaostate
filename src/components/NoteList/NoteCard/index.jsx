import React, {useContext} from "react";
import { NotesContext } from "../../../contexts/NotesContext";

const NoteCard = ({note}) => {
  const { removeNote, setCurrentNote } = useContext(NotesContext);

  return (
    <li key={note.id}>
      <h2>{note.title}</h2>
      <p>{note.text}</p>
      <button onClick={() => removeNote(note.id)}>Remover</button>
      <button onClick={() => setCurrentNote(note)}>Editar</button>
    </li>
  );
};

export default NoteCard;
