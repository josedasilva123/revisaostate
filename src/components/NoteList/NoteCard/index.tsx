import React, {useContext} from "react";
import { INote, NotesContext } from "../../../contexts/NotesContext";
import { StyledNoteCard } from "./style";
import { MdDelete, MdEdit} from "react-icons/md";

interface INoteCardsProps{
  note: INote;
}

const NoteCard = ({note}: INoteCardsProps) => {
  const { removeNote, setCurrentNote } = useContext(NotesContext);

  return (
    <StyledNoteCard>
      <div className="content">
        <h2>{note.title}</h2>
        <p>{note.text}</p>
      </div>
      <div className="controls">
      <button onClick={() => removeNote(note._id)}>
        <MdDelete size={24} />
      </button>
      <button onClick={() => setCurrentNote(note)}>
        <MdEdit size={24} />
      </button>
      </div>      
    </StyledNoteCard>
  );
};

export default NoteCard;
