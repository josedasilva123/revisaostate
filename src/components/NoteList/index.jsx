import React from "react";
import NoteCard from "./NoteCard";

const NoteList = ({ notes,  handleRemove, selectNote }) => {
  return (
    <div>
      {notes.map((note) => (
        <NoteCard key={note.id} note={note} handleRemove={handleRemove} selectNote={selectNote} />
      ))}
    </div>
  );
};

export default NoteList;
