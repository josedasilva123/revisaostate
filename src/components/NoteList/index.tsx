import React, {useContext} from "react";
import { NotesContext } from "../../contexts/NotesContext";
import NoteCard from "./NoteCard";
import { StyledNoteList } from "./style";

const NoteList = () => {
  const { notes } = useContext(NotesContext);
  return (
    <StyledNoteList>
      {notes.length ? notes?.map((note) => (
        <NoteCard key={note._id} note={note} />
      )) : (
        <h1>Nenhuma nota cadastrada. Cadastre uma nova nota</h1>
      )}
    </StyledNoteList>
  );
};

export default NoteList;
