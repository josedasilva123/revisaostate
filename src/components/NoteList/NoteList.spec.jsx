/* eslint-disable testing-library/no-debugging-utils */
import { render, screen } from "@testing-library/react";
import NoteList from ".";
import { NotesContext } from "../../contexts/NotesContext";

describe("<NoteList />", () => {
  it("should render card list", () => {
    const providerProps = {
      notes: [
        {
          _id: "123456",
          userID: "654321",
          title: "Nota 1",
          text: "Está é a nota número 1",
          createdAt: "Data de criação",
          updatedAt: "Data de atualização",
          __v: 1,
        },
        {
          _id: "654321",
          userID: "654321",
          title: "Nota 2",
          text: "Está é a nota número 2",
          createdAt: "Data de criação",
          updatedAt: "Data de atualização",
          __v: 1,
        },
      ],
    };

    render(
      <NotesContext.Provider value={providerProps}>
        <NoteList />
      </NotesContext.Provider>
    );

    expect(screen.getByText("Nota 1")).toBeInTheDocument();
    expect(screen.getByText("Nota 2")).toBeInTheDocument();
    expect(screen.getAllByRole('heading')).toHaveLength(2);
  });

  it("should render no note message and notes is empty", () => {
    const providerProps = {
      notes: [],
    };

    render(
      <NotesContext.Provider value={providerProps}>
        <NoteList />
      </NotesContext.Provider>
    );

    expect(screen.getByText(/Nenhuma nota cadastrada/i)).toBeInTheDocument();

  });

  it("should match snapshot", () => {
    const providerProps = {
      notes: [
        {
          _id: "123456",
          userID: "654321",
          title: "Nota 1",
          text: "Está é a nota número 1",
          createdAt: "Data de criação",
          updatedAt: "Data de atualização",
          __v: 1,
        },
        {
          _id: "654321",
          userID: "654321",
          title: "Nota 2",
          text: "Está é a nota número 2",
          createdAt: "Data de criação",
          updatedAt: "Data de atualização",
          __v: 1,
        },
      ],
    };
    
    const { container } = render(
      <NotesContext.Provider value={providerProps}>
        <NoteList />
      </NotesContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
