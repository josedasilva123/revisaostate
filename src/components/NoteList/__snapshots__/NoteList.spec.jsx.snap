// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteList /> should match snapshot 1`] = `
<div>
  <ul
    class="sc-gsnTZi dQOFoV"
  >
    <li
      class="sc-bczRLJ DjLSX"
    >
      <div
        class="content"
      >
        <h2>
          Nota 1
        </h2>
        <p>
          Está é a nota número 1
        </p>
      </div>
      <div
        class="controls"
      >
        <button
          title="Delete"
        >
          <svg
            fill="currentColor"
            height="24"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </button>
      </div>
    </li>
    <li
      class="sc-bczRLJ DjLSX"
    >
      <div
        class="content"
      >
        <h2>
          Nota 2
        </h2>
        <p>
          Está é a nota número 2
        </p>
      </div>
      <div
        class="controls"
      >
        <button
          title="Delete"
        >
          <svg
            fill="currentColor"
            height="24"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </button>
      </div>
    </li>
  </ul>
</div>
`;
